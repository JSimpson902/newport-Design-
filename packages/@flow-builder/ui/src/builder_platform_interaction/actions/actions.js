import { isCanvasElement } from "builder_platform_interaction/elementConfig";
import { ELEMENT_TYPE } from "builder_platform_interaction/flowMetadata";
import { UNDO, REDO } from "builder_platform_interaction/undoRedoLib";

export const UPDATE_FLOW = 'UPDATE_FLOW';

export const UPDATE_PROPERTIES = 'UPDATE_PROPERTIES';
export const UPDATE_PROPERTIES_AFTER_SAVING = 'UPDATE_PROPERTIES_AFTER_SAVING';

export const ADD_CANVAS_ELEMENT = 'ADD_CANVAS_ELEMENT';
export const UPDATE_CANVAS_ELEMENT = 'UPDATE_CANVAS_ELEMENT';

export const DELETE_ELEMENT = 'DELETE_ELEMENT';

export const ADD_RESOURCE = 'ADD_RESOURCE';
export const UPDATE_VARIABLE_CONSTANT = 'UPDATE_VARIABLE_CONSTANT';
export const UPDATE_RESOURCE = 'UPDATE_RESOURCE';
export const DELETE_RESOURCE = 'DELETE_RESOURCE';

export const ADD_CONNECTOR = 'ADD_CONNECTOR';

export const SELECT_ON_CANVAS = 'SELECT_ON_CANVAS';
export const TOGGLE_ON_CANVAS = 'TOGGLE_ON_CANVAS';
export const DESELECT_ON_CANVAS = 'DESELECT_ON_CANVAS';

export const HIGHLIGHT_ON_CANVAS = 'HIGHLIGHT_ON_CANVAS';
export const UNHIGHLIGHT_ON_CANVAS = 'UNHIGHLIGHT_ON_CANVAS';

export const ADD_DECISION_WITH_OUTCOMES = 'ADD_DECISION_WITH_OUTCOMES';
export const MODIFY_DECISION_WITH_OUTCOMES = 'MODIFY_DECISION_WITH_OUTCOMES';

export const ADD_WAIT_WITH_WAIT_EVENTS = 'ADD_WAIT_WITH_WAIT_EVENTS';
export const MODIFY_WAIT_WITH_WAIT_EVENTS = 'MODIFY_WAIT_WITH_WAIT_EVENTS';

export const UPDATE_RECORD_LOOKUP = 'UPDATE_RECORD_LOOKUP';

export const ADD_SCREEN_WITH_FIELDS = 'ADD_SCREEN_WITH_FIELDS';
export const MODIFY_SCREEN_WITH_FIELDS = 'MODIFY_SCREEN_WITH_FIELDS';

export const ADD_START_ELEMENT = 'ADD_START_ELEMENT';


export const PROPERTY_EDITOR_ACTION = {
    UPDATE_ELEMENT_PROPERTY: 'UPDATE_ELEMENT_PROPERTY',
    ADD_ASSIGNMENT_ITEM: 'ADD_ASSIGNMENT_ITEM',
    DELETE_ASSIGNMENT_ITEM: 'DELETE_ASSIGNMENT_ITEM',
    UPDATE_ASSIGNMENT_ITEM: 'UPDATE_ASSIGNMENT_ITEM',
    CHANGE_ACTION_TYPE: 'CHANGE_ACTION_TYPE',
    ADD_DECISION_OUTCOME: 'ADD_DECISION_OUTCOME',
    ADD_WAIT_EVENT: 'ADD_WAIT_EVENT',
    CHANGE_DATA_TYPE: 'CHANGE_DATA_TYPE',
    UPDATE_ELEMENT_VALUE: 'UPDATE_ELEMENT_VALUE',
    ADD_SCREEN_FIELD: 'ADD_SCREEN_FIELD',
    UPDATE_FILTER_ITEM:'updaterecordfilter',
    DELETE_FILTER_ITEM: 'deleterecordfilter',
    ADD_FILTER_ITEM: 'addrecordfilter',
    ADD_EMPTY_OUTPUT_ASSIGNMENT: 'addemptyoutputassignment',
    UPDATE_OUTPUT_ASSIGNMENTS_BEFORE_CLOSE: 'updateoutputassignmentbeforeclose'
};

/**
 * Helper function to create actions.
 *
 * @param {String} type - action type
 * @param {Object} payload - data
 * @returns {Object} action new action based on type and payload
 */
export const createAction = (type, payload = {}) => ({type, payload});

/**
 * Action for updating flow information in the store.
 * To be used when we get flow metadata from backend or when we create a new flow.
 *
 * @param {Object} payload - contains new flow information
 * @returns {Object} action new action based on type and payload
 */
export const updateFlow = (payload) => createAction(UPDATE_FLOW, payload);

/**
 * @returns {Object} Undo Action without any payload.
 */
export const undo = createAction(UNDO);

/**
 * @returns {Object} Redo Action without any payload.
 */
export const redo = createAction(REDO);

/**
 * Action for updating flow properties in the store.
 *
 * @param {Object} payload - contains new flow information
 * @returns {Object} action new action based on type and payload
 */
export const updateProperties = (payload) => createAction(UPDATE_PROPERTIES, payload);

/**
 * Action for updating flow properties in the store after saving a flow.
 *
 * @param {Object} payload - contains new flow information
 * @returns {Object} action new action based on type and payload
 */
export const updatePropertiesAfterSaving = (payload) => createAction(UPDATE_PROPERTIES_AFTER_SAVING, payload);

/**
 * Action for adding a new element in the store.
 *
 * @param {Object} payload - contains new element value to be added
 * @returns {Object} action new action based on type and payload
 */
export const addElement = (payload) => {
    if (payload) {
        switch (payload.elementType) {
            case ELEMENT_TYPE.VARIABLE:
            case ELEMENT_TYPE.CONSTANT:
            case ELEMENT_TYPE.FORMULA:
            case ELEMENT_TYPE.TEXT_TEMPLATE:
            case ELEMENT_TYPE.STAGE:
            case ELEMENT_TYPE.CHOICE:
            case ELEMENT_TYPE.PICKLIST_CHOICE_SET:
            case ELEMENT_TYPE.RECORD_CHOICE_SET:
                return createAction(ADD_RESOURCE, payload);
            case ELEMENT_TYPE.DECISION_WITH_MODIFIED_AND_DELETED_OUTCOMES:
                return createAction(ADD_DECISION_WITH_OUTCOMES, payload);
            case ELEMENT_TYPE.WAIT_WITH_MODIFIED_AND_DELETED_WAIT_EVENTS:
                return createAction(ADD_WAIT_WITH_WAIT_EVENTS, payload);
            case ELEMENT_TYPE.SCREEN_WITH_MODIFIED_AND_DELETED_SCREEN_FIELDS:
                return createAction(ADD_SCREEN_WITH_FIELDS, payload);
            case ELEMENT_TYPE.START_ELEMENT:
                return createAction(ADD_START_ELEMENT, payload);
            default:
                if (isCanvasElement(payload.elementType)) {
                    return createAction(ADD_CANVAS_ELEMENT, payload);
                }

                // Added to support strategy builder non-canvas elements
                return createAction(ADD_RESOURCE, payload);
        }
    }
    return {};
};

/**
 * Action for updating an element in the store.
 *
 * For variables & constants, all existing data for the element in the store will be overwritten
 *
 * @param {Object} payload - contains GUID of the element to be updated and new values
 * @returns {Object} action new action based on type and payload
 */
export const updateElement = (payload) => {
    if (payload) {
        switch (payload.elementType) {
            case ELEMENT_TYPE.VARIABLE:
            case ELEMENT_TYPE.CONSTANT:
                return createAction(UPDATE_VARIABLE_CONSTANT, payload);
            case ELEMENT_TYPE.TEXT_TEMPLATE:
            case ELEMENT_TYPE.FORMULA:
            case ELEMENT_TYPE.STAGE:
            case ELEMENT_TYPE.CHOICE:
            case ELEMENT_TYPE.PICKLIST_CHOICE_SET:
            case ELEMENT_TYPE.RECORD_CHOICE_SET:
                return createAction(UPDATE_RESOURCE, payload);
            case ELEMENT_TYPE.DECISION_WITH_MODIFIED_AND_DELETED_OUTCOMES:
                return createAction(MODIFY_DECISION_WITH_OUTCOMES, payload);
            case ELEMENT_TYPE.WAIT_WITH_MODIFIED_AND_DELETED_WAIT_EVENTS:
                return createAction(MODIFY_WAIT_WITH_WAIT_EVENTS, payload);
            case ELEMENT_TYPE.RECORD_LOOKUP:
                return createAction(UPDATE_RECORD_LOOKUP, payload);
            case ELEMENT_TYPE.SCREEN_WITH_MODIFIED_AND_DELETED_SCREEN_FIELDS:
                return createAction(MODIFY_SCREEN_WITH_FIELDS, payload);
            default:
                if (isCanvasElement(payload.elementType)) {
                    return createAction(UPDATE_CANVAS_ELEMENT, payload);
                }

                // Added to support strategy builder non-canvas elements
                return createAction(UPDATE_RESOURCE, payload);
        }
    }
    return {};
};

/**
 * Action for deleting an element from the store.
 *
 * @param {Object} payload - contains GUID of the element to be deleted
 * @returns {Object} action new action based on type and payload
 */
export const deleteElement = (payload) => {
    if (!payload) {
        return {};
    }

    let action;
    switch (payload.elementType) {
        case ELEMENT_TYPE.VARIABLE:
        case ELEMENT_TYPE.CONSTANT:
        case ELEMENT_TYPE.TEXT_TEMPLATE:
        case ELEMENT_TYPE.STAGE:
        case ELEMENT_TYPE.CHOICE:
        case ELEMENT_TYPE.PICKLIST_CHOICE_SET:
        case ELEMENT_TYPE.RECORD_CHOICE_SET:
        case ELEMENT_TYPE.FORMULA:
            action = createAction(DELETE_RESOURCE, payload);
            break;
        default:
            // Check if element is a canvas element or multiple elements are being deleted
            if (isCanvasElement(payload.elementType) || !payload.elementType) {
                action = createAction(DELETE_ELEMENT, payload);
            } else {
                // Added to support strategy builder non-canvas elements
                action = createAction(DELETE_RESOURCE, payload);
            }
    }

    return action;
};

/**
 * Action for adding a new connector to the store.
 *
 * @param {Object} payload - contains connectorGUID, source, target, connector label and config
 * @returns {Object} action new action based on type and payload
 */
export const addConnector = (payload) => createAction(ADD_CONNECTOR, payload);

/**
 * Action for selecting a canvas element or connector.
 *
 * @param {Object} payload - contains GUID of the element to be selected
 * @returns {Object} action new action based on type and payload
 */
export const selectOnCanvas = (payload) => createAction(SELECT_ON_CANVAS, payload);

/**
 * Action for toggling the isSelected property of a canvas element or connector.
 *
 * @param {Object} payload - contains GUID of the element to be toggled
 * @returns {Object} action new action based on type and payload
 */
export const toggleOnCanvas = (payload) => createAction(TOGGLE_ON_CANVAS, payload);

/**
 * Action for deselecting all the selected canvas elements and connectors.
 *
 * @returns {Object} action new action based on type and payload
 */
export const deselectOnCanvas = createAction(DESELECT_ON_CANVAS);

/**
 * Action for setting the isHighlighted property of a canvas element to true
 *
 * @param {Object} payload - contains GUID of the element to be toggled
 * @returns {Object} action new action based on type and payload
 */
export const highlightOnCanvas = (payload) => createAction(HIGHLIGHT_ON_CANVAS, payload);

/**
 * Action for setting the isHighlighted property of a canvas element to false
 *
 * @param {Object} payload - contains GUID of the element to be toggled
 * @returns {Object} action new action based on type and payload
 */
export const unhighlightOnCanvas = (payload) => createAction(UNHIGHLIGHT_ON_CANVAS, payload);