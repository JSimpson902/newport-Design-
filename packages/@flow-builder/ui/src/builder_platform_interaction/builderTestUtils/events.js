export const focusEvent = new FocusEvent('focus', {
    bubbles: false,
    cancelable: false
});

export const focusoutEvent = new FocusEvent('focusout', {
    bubbles: true,
    cancelable: true
});

export const textInputEvent = textInput => {
    return new CustomEvent('textinput', {
        bubbles: true,
        cancelable: true,
        detail: { text: textInput }
    });
};

export const blurEvent = new FocusEvent('blur', {
    bubbles: true,
    cancelable: true
});

export const selectEvent = value => {
    return new CustomEvent('select', {
        bubbles: true,
        cancelable: true,
        detail: { value }
    });
};

export const changeEvent = value => {
    return new CustomEvent('change', {
        bubbles: true,
        cancelable: false,
        detail: { value }
    });
};

export const clickEvent = () => {
    return new CustomEvent('click', {
        bubbles: true,
        cancelable: false
    });
};

export const checkboxChangeEvent = (checked = true) => {
    return new CustomEvent('change', {
        bubbles: true,
        cancelable: false,
        detail: { checked }
    });
};
