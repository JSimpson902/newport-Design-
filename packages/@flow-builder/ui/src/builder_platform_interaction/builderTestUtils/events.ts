// @ts-nocheck
export const focusEvent = new FocusEvent('focus', {
    bubbles: false,
    cancelable: false
});

export const focusoutEvent = new FocusEvent('focusout', {
    bubbles: true,
    cancelable: true
});

export const textInputEvent = textInput => {
    return new CustomEvent('textinput', {
        bubbles: true,
        cancelable: true,
        detail: { text: textInput }
    });
};

export const blurEvent = new FocusEvent('blur', {
    bubbles: true,
    cancelable: true,
    composed: true
});

export const selectEvent = value => {
    return new CustomEvent('select', {
        bubbles: true,
        cancelable: true,
        detail: { value }
    });
};

export const changeEvent = value => {
    return new CustomEvent('change', {
        bubbles: true,
        cancelable: false,
        detail: { value }
    });
};

export const clickEvent = () => {
    return new CustomEvent('click', {
        bubbles: true,
        cancelable: false
    });
};

export const mouseoverEvent = () => {
    return new CustomEvent('mouseover', {
        bubbles: true,
        cancelable: false
    });
};

export const mouseoutEvent = () => {
    return new CustomEvent('mouseout', {
        bubbles: true,
        cancelable: false
    });
};

export const checkboxChangeEvent = (checked = true) => {
    return new CustomEvent('change', {
        bubbles: true,
        cancelable: false,
        detail: { checked }
    });
};

/**
 * Remove event (used on pill for instance)
 */
export const removeEvent = (): CustomEvent => {
    return new CustomEvent('remove', {
        bubbles: true,
        cancelable: false
    });
};

export const lightningRadioGroupChangeEvent = newValue => {
    return new CustomEvent('change', {
        detail: {
            newValue
        },
        composed: true,
        bubbles: true,
        cancelable: true
    });
};
