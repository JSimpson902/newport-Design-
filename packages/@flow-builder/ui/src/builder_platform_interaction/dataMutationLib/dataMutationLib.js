export * from './arrayMutation';
export * from './objectMutation';
export * from './elementDataMutation';
export * from './stringMutation';
export * from './screenEditorDataMutation';
export * from './mergeFieldsMutation';
export { GUID_SUFFIX, mutateFEROV, deMutateFEROV } from './ferovEditorDataMutation';