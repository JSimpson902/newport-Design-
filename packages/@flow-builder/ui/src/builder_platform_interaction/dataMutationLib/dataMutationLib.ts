export * from './arrayMutation';
export * from './objectMutation';
export * from './elementDataMutation';
export * from './stringMutation';
export * from './mergeFieldsMutation';
