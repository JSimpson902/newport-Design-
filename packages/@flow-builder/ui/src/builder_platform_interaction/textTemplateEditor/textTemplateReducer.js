import { updateProperties } from 'builder_platform_interaction/dataMutationLib';
import { PROPERTY_EDITOR_ACTION } from 'builder_platform_interaction/actions';
import { VALIDATE_ALL } from 'builder_platform_interaction/validationRules';
import { textTemplateValidation } from './textTemplateValidation';

const textTemplatePropertyChanged = (textTemplate, action) => {
    action.payload.error =
        action.payload.error === null
            ? textTemplateValidation.validateProperty(
                  action.payload.propertyName,
                  action.payload.value
              )
            : action.payload.error;
    return updateProperties(textTemplate, {
        [action.payload.propertyName]: {
            error: action.payload.error,
            value: action.payload.value
        }
    });
};

/**
 * Text template reducer function runs validation rules and returns back the updated variable
 * @param {object} textTemplate - element / node state
 * @param {object} action - object containing type and payload eg: {type:"xyz", payload: {propertyName: '', value: '' , error: ''}}
 * @returns {object} textTemplate - updated text template
 */
export const textTemplateReducer = (textTemplate, action) => {
    switch (action.type) {
        case PROPERTY_EDITOR_ACTION.UPDATE_ELEMENT_PROPERTY:
            return textTemplatePropertyChanged(textTemplate, action);
        case VALIDATE_ALL:
            return textTemplateValidation.validateAll(textTemplate);
        default:
            return textTemplate;
    }
};
