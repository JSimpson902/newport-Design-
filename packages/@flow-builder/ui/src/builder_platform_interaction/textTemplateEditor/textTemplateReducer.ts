// @ts-nocheck
import { updateProperties } from 'builder_platform_interaction/dataMutationLib';
import { PROPERTY_EDITOR_ACTION } from 'builder_platform_interaction/actions';
import { VALIDATE_ALL } from 'builder_platform_interaction/validationRules';
import { textTemplateValidation } from './textTemplateValidation';
import { ELEMENT_TYPE } from 'builder_platform_interaction/flowMetadata';
import { elementTypeToConfigMap } from 'builder_platform_interaction/elementConfig';

const NON_HYDRATABLE_PROPS = new Set([...elementTypeToConfigMap[ELEMENT_TYPE.TEXT_TEMPLATE].nonHydratableProperties]);

const textTemplatePropertyChanged = (textTemplate, action) => {
    const isHydrated = !NON_HYDRATABLE_PROPS.has(action.payload.propertyName);
    action.payload.error =
        action.payload.error === null
            ? textTemplateValidation.validateProperty(action.payload.propertyName, action.payload.value)
            : action.payload.error;
    return updateProperties(textTemplate, {
        [action.payload.propertyName]: isHydrated
            ? {
                  error: action.payload.error,
                  value: action.payload.value
              }
            : action.payload.value
    });
};

/**
 * Text template reducer function runs validation rules and returns back the updated variable
 *
 * @param {object} textTemplate - element / node state
 * @param {object} action - object containing type and payload eg: {type:"xyz", payload: {propertyName: '', value: '' , error: ''}}
 * @returns {object} textTemplate - updated text template
 */
export const textTemplateReducer = (textTemplate, action) => {
    switch (action.type) {
        case PROPERTY_EDITOR_ACTION.UPDATE_ELEMENT_PROPERTY:
            return textTemplatePropertyChanged(textTemplate, action);
        case VALIDATE_ALL:
            return textTemplateValidation.validateAll(textTemplate);
        default:
            return textTemplate;
    }
};
