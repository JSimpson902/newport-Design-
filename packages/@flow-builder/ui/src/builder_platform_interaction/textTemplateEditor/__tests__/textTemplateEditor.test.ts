// @ts-nocheck
import { createElement } from 'lwc';
import TextTemplateEditor from '../textTemplateEditor';
import { textTemplateReducer } from '../textTemplateReducer';
import * as mockStoreData from 'mock/storeData';
import { createAction, PROPERTY_EDITOR_ACTION } from 'builder_platform_interaction/actions';
import { PropertyChangedEvent } from 'builder_platform_interaction/events';
import { VALIDATE_ALL } from 'builder_platform_interaction/validationRules';
import { getErrorsFromHydratedElement } from 'builder_platform_interaction/dataMutationLib';
import { ELEMENT_TYPE } from 'builder_platform_interaction/flowMetadata';
import {
    ticks,
    INTERACTION_COMPONENTS_SELECTORS,
    LIGHTNING_COMPONENTS_SELECTORS
} from 'builder_platform_interaction/builderTestUtils';

jest.mock('builder_platform_interaction/ferovResourcePicker', () =>
    require('builder_platform_interaction_mocks/ferovResourcePicker')
);

const SELECTORS = {
    ...INTERACTION_COMPONENTS_SELECTORS,
    ...LIGHTNING_COMPONENTS_SELECTORS
};

const setupComponentUnderTest = (props) => {
    const element = createElement('builder_platform_interaction-text-template-editor', {
        is: TextTemplateEditor
    });
    element.node = props;
    document.body.appendChild(element);
    return element;
};

jest.mock('builder_platform_interaction/dataMutationLib', () => {
    const actual = jest.requireActual('builder_platform_interaction/dataMutationLib');
    return {
        pick: actual.pick,
        getErrorsFromHydratedElement: jest.fn(),
        getValueFromHydratedItem: actual.getValueFromHydratedItem
    };
});

jest.mock('builder_platform_interaction/actions', () => {
    return {
        createAction: jest.fn().mockImplementation((type, payload) => payload),
        PROPERTY_EDITOR_ACTION: jest.requireActual('builder_platform_interaction/actions').PROPERTY_EDITOR_ACTION
    };
});

jest.mock('../textTemplateReducer', () => {
    return {
        textTemplateReducer: jest.fn().mockImplementation((obj) => Object.assign({}, obj))
    };
});

const mockHydratedElementWithErrors = [{ key: 'mockKey', errorString: 'mockErrorString' }];

describe('text-template-editor', () => {
    let textTemplateInRichTextModeResource, textTemplateInPlainTextModeResource;

    beforeEach(() => {
        textTemplateInRichTextModeResource = mockStoreData.textTemplateInRichTextModeForPropertyEditor();
        textTemplateInPlainTextModeResource = mockStoreData.textTemplateInPlainTextModeForPropertyEditor();
    });

    test('check UI (snapshot) - "RichTextPlainTextSwitch" displayed with "richText" mode checked by default, and correct css class for resource picker', () => {
        const textTemplateEditor = setupComponentUnderTest(textTemplateInRichTextModeResource);
        expect(textTemplateEditor).toMatchSnapshot();
    });

    test('check UI (snapshot) - "RichTextPlainTextSwitch" displayed with "richText" mode off', () => {
        const textTemplateEditor = setupComponentUnderTest(textTemplateInPlainTextModeResource);
        expect(textTemplateEditor).toMatchSnapshot();
    });

    it('contains a text template element', async () => {
        const textTemplateEditor = setupComponentUnderTest(textTemplateInRichTextModeResource);
        await ticks(1);
        expect(textTemplateEditor.node.elementType).toEqual(ELEMENT_TYPE.TEXT_TEMPLATE);
        expect(textTemplateEditor.getNode()).toEqual(textTemplateInRichTextModeResource);
    });

    it('has label description component that shows only dev name and description', () => {
        const textTemplateEditor = setupComponentUnderTest(textTemplateInRichTextModeResource);
        const labelDescription = textTemplateEditor.shadowRoot.querySelector(SELECTORS.LABEL_DESCRIPTION);
        expect(labelDescription).toBeDefined();
        expect(labelDescription.description).toEqual(textTemplateInRichTextModeResource.description);
        expect(labelDescription.devName).toEqual(textTemplateInRichTextModeResource.name);
        expect(labelDescription.hideLabel).toBeTruthy();
    });

    it('has resourced-textarea showing the correct text template', () => {
        const textTemplateEditor = setupComponentUnderTest(textTemplateInRichTextModeResource);
        const resourcedRichTextEditor = textTemplateEditor.shadowRoot.querySelector(
            SELECTORS.RESOURCED_RICH_TEXT_EDITOR
        );
        expect(resourcedRichTextEditor).toBeDefined();
        const lightningInputRichText = resourcedRichTextEditor.shadowRoot.querySelector(
            SELECTORS.LIGHTNING_INPUT_RICH_TEXT
        );
        expect(lightningInputRichText.value).toEqual(textTemplateInRichTextModeResource.text.value);
    });

    it('handles the property changed event and updates the property from label-description', async () => {
        const textTemplateEditor = setupComponentUnderTest(textTemplateInRichTextModeResource);
        await ticks(1);
        const event = new PropertyChangedEvent('description', 'new desc', null);
        textTemplateEditor.shadowRoot
            .querySelector('builder_platform_interaction-label-description')
            .dispatchEvent(event);
        expect(createAction.mock.calls[0][0]).toEqual(PROPERTY_EDITOR_ACTION.UPDATE_ELEMENT_PROPERTY);
        expect(textTemplateReducer.mock.calls[0][0]).toEqual(textTemplateEditor.node);
    });

    it('handles the change event and updates the property from resourced-textarea', async () => {
        const textTemplateEditor = setupComponentUnderTest(textTemplateInRichTextModeResource);
        const newTextTemplate = '<html> New text in template</html>';
        await ticks(1);
        const event = new CustomEvent('change', {
            detail: { value: newTextTemplate, error: null },
            cancelable: true,
            composed: true,
            bubbles: true
        });
        textTemplateEditor.shadowRoot.querySelector(SELECTORS.RESOURCED_RICH_TEXT_EDITOR).dispatchEvent(event);
        expect(createAction.mock.calls[0][0]).toEqual(PROPERTY_EDITOR_ACTION.UPDATE_ELEMENT_PROPERTY);
        expect(createAction.mock.calls[0][1]).toEqual({
            propertyName: 'text',
            value: newTextTemplate,
            error: null
        });
        expect(textTemplateReducer.mock.calls[0][0]).toEqual(textTemplateEditor.node);
    });

    it('handles the text mode switch event holding mode plain text and updates the property isPlainTextMode', async () => {
        const textTemplateEditor = setupComponentUnderTest(textTemplateInRichTextModeResource);
        await ticks(1);
        const event = new CustomEvent('richtextplaintextswitchchanged', {
            detail: { isPlainText: true },
            cancelable: true,
            composed: true,
            bubbles: true
        });
        textTemplateEditor.shadowRoot.querySelector(SELECTORS.RESOURCED_RICH_TEXT_EDITOR).dispatchEvent(event);
        expect(createAction.mock.calls[0][0]).toEqual(PROPERTY_EDITOR_ACTION.UPDATE_ELEMENT_PROPERTY);
        expect(createAction.mock.calls[0][1]).toEqual({
            propertyName: 'isViewedAsPlainText',
            value: true
        });
        expect(textTemplateReducer.mock.calls[0][0]).toEqual(textTemplateEditor.node);
    });

    it('handles the text mode switch event holding mode rich text and updates the property isPlainTextMode', async () => {
        const textTemplateEditor = setupComponentUnderTest(textTemplateInPlainTextModeResource);
        await ticks(1);
        const event = new CustomEvent('richtextplaintextswitchchanged', {
            detail: { isPlainText: false },
            cancelable: true,
            composed: true,
            bubbles: true
        });
        textTemplateEditor.shadowRoot.querySelector(SELECTORS.RESOURCED_RICH_TEXT_EDITOR).dispatchEvent(event);
        expect(createAction.mock.calls[0][0]).toEqual(PROPERTY_EDITOR_ACTION.UPDATE_ELEMENT_PROPERTY);
        expect(createAction.mock.calls[0][1]).toEqual({
            propertyName: 'isViewedAsPlainText',
            value: false
        });
        expect(textTemplateReducer.mock.calls[0][0]).toEqual(textTemplateEditor.node);
    });

    describe('validation', () => {
        it('calls reducer with validate all event', () => {
            const textTemplateEditor = setupComponentUnderTest(textTemplateInRichTextModeResource);
            const node = textTemplateEditor.node;
            textTemplateEditor.validate();
            expect(textTemplateReducer.mock.calls[0][0]).toEqual(node);
            expect(textTemplateReducer.mock.calls[0][1]).toEqual({
                type: VALIDATE_ALL
            });
        });

        it('gets the errors after validating', () => {
            const textTemplateEditor = setupComponentUnderTest(textTemplateInRichTextModeResource);
            getErrorsFromHydratedElement.mockReturnValueOnce(mockHydratedElementWithErrors);
            const result = textTemplateEditor.validate();
            expect(getErrorsFromHydratedElement).toHaveBeenCalledWith(textTemplateEditor.node);
            expect(result).toEqual(mockHydratedElementWithErrors);
        });
    });
});
