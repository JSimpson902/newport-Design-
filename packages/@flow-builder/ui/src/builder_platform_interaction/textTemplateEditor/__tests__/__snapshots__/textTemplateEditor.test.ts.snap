// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text-template-editor check UI (snapshot) - "RichTextPlainTextSwitch" displayed with "richText" mode checked by default, and correct css class for resource picker 1`] = `
<builder_platform_interaction-text-template-editor>
  #shadow-root(open)
    <div
      class="slds-form slds-form_compound slds-p-around_small"
    >
      <builder_platform_interaction-label-description>
        #shadow-root(open)
          <span
            class="slds-hide"
          >
            HACK TO FORCE RERENDER:
          </span>
          <div>
            <div
              class="slds-form slds-form_compound"
            >
              <div
                class="container slds-form-element__group slds-grid slds-gutters slds-gutters_xx-small slds-m-bottom_x-small"
              >
                <lightning-input
                  class="slds-col devName"
                >
                  #shadow-root(open)
                </lightning-input>
              </div>
              <div
                class="slds-form-element__group slds-form-element__row slds-p-bottom_xx-small"
              >
                <lightning-textarea
                  class="slds-size_1-of-1 description"
                >
                  #shadow-root(open)
                </lightning-textarea>
              </div>
            </div>
          </div>
      </builder_platform_interaction-label-description>
      <builder_platform_interaction-resourced-rich-text-editor>
        #shadow-root(open)
          <div
            class="property-input slds-form-element"
          >
            <div
              class="slds-grid slds-grid_vertical"
            >
              <div
                class="slds-col slds-order_1 slds-size_1-of-1"
              >
                <span
                  class="slds-form-element__label"
                  id="[shadow:guid]"
                >
                  <abbr
                    class="slds-required"
                    title="FlowBuilderResourcedRichTextEditor.inputRequiredTitle"
                  >
                    *
                  </abbr>
                  FlowBuilderTextTemplateEditor.textTemplateLabel
                </span>
                <lightning-helptext
                  class="slds-form-element"
                >
                  #shadow-root(open)
                </lightning-helptext>
              </div>
            </div>
            <div
              aria-labelledby="[shadow:guid]"
              class="container slds-grid slds-grid_vertical"
            >
              <div>
                <div
                  class="divParentResourcePickerModeOptions slds-p-around_x-small slds-p-bottom_xx-small slds-button-group-list"
                >
                  <div
                    class="divResourcePickerPartialWidth"
                  >
                    <builder_platform_interaction-ferov-resource-picker>
                      #shadow-root(open)
                    </builder_platform_interaction-ferov-resource-picker>
                  </div>
                  <builder_platform_interaction-rich-text-plain-text-switch>
                    #shadow-root(open)
                      <div
                        class="divPlainRichTextSwitch"
                      >
                        <lightning-combobox>
                          #shadow-root(open)
                        </lightning-combobox>
                      </div>
                  </builder_platform_interaction-rich-text-plain-text-switch>
                </div>
                <lightning-input-rich-text>
                  #shadow-root(open)
                </lightning-input-rich-text>
              </div>
            </div>
          </div>
      </builder_platform_interaction-resourced-rich-text-editor>
    </div>
</builder_platform_interaction-text-template-editor>
`;
